File: hash_dag.py
"""Bidirectional graph construction across hash nodes by shared input bits."""

from __future__ import annotations

from dataclasses import dataclass

from group_hash import HashNode


@dataclass(frozen=True)
class GraphEdge:
    src: str
    dst: str
    weight: int


@dataclass
class HashGraph:
    nodes: dict[str, HashNode]
    edges: list[GraphEdge]
    adj: dict[str, list[GraphEdge]]


def build_hash_graph(nodes: list[HashNode]) -> HashGraph:
    node_map = {node.node_id: node for node in nodes}
    edges: list[GraphEdge] = []
    adj: dict[str, list[GraphEdge]] = {node.node_id: [] for node in nodes}

    for i in range(len(nodes)):
        for j in range(i + 1, len(nodes)):
            left = nodes[i]
            right = nodes[j]
            overlap = len(left.source_indices.intersection(right.source_indices))
            if overlap <= 0:
                continue
            edge = GraphEdge(src=left.node_id, dst=right.node_id, weight=overlap)
            edges.append(edge)
            adj[left.node_id].append(edge)
            adj[right.node_id].append(edge)

    return HashGraph(nodes=node_map, edges=edges, adj=adj)

File: cpp/src/hash_dag.cpp
#include "hash_dag.hpp"
#include <algorithm>

HashGraph build_hash_graph(const std::vector<HashNode>& nodes) {
    std::unordered_map<std::string, HashNode> node_map;
    std::unordered_map<std::string, std::vector<GraphEdge>> adj;
    std::vector<GraphEdge> edges;

    node_map.reserve(nodes.size());
    adj.reserve(nodes.size());
    for (const auto& node : nodes) {
        node_map[node.node_id] = node;
        adj[node.node_id] = {};
    }

    for (int i = 0; i < (int)nodes.size(); i++) {
        for (int j = i + 1; j < (int)nodes.size(); j++) {
            const auto& a = nodes[i];
            const auto& b = nodes[j];

            // Both source_indices vectors are sorted; use set_intersection
            std::vector<int> isect;
            std::set_intersection(
                a.source_indices.begin(), a.source_indices.end(),
                b.source_indices.begin(), b.source_indices.end(),
                std::back_inserter(isect));

            int overlap = (int)isect.size();
            if (overlap <= 0) continue;

            GraphEdge edge{a.node_id, b.node_id, overlap};
            edges.push_back(edge);
            adj[a.node_id].push_back(edge);
            adj[b.node_id].push_back(edge);
        }
    }

    return HashGraph{std::move(node_map), std::move(edges), std::move(adj)};
}
